torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,18 +10,56 @@
 #include <cmath>
 #include <functional>
 #include <vector>
+#include <algorithm>
 
 namespace torch {
 namespace optim {
 
-LBFGSOptions::LBFGSOptions(double learning_rate)
-    : learning_rate_(learning_rate) {}
+LBFGSOptions::LBFGSOptions(double lr) : lr_(lr) {}
 
-Tensor LBFGS::gather_flat_grad() {
+// TORCH_ARG(double, lr) = 1;
+// TORCH_ARG(int64_t, max_iter) = 20;
+// TORCH_ARG(c10::optional<int64_t>, max_eval) = c10::nullopt;
+// TORCH_ARG(double, tolerance_grad) = 1e-7;
+// TORCH_ARG(double, tolerance_change) = 1e-9;
+// TORCH_ARG(size_t, history_size) = 100;
+// TORCH_ARG(c10::optional<std::string>, line_search_fn) = c10::nullopt;
+bool operator==(const LBFGSOptions& lhs, const LBFGSOptions& rhs) {
+  auto isNull = [](c10::optional<int64_t>& max_eval){ return max_eval == c10::nullopt; };
+  return (lhs.lr() == rhs.lr()) &&
+         (lhs.max_iter() == rhs.max_iter()) &&
+         // ((isNull(lhs.max_eval()) && isNull(rhs.max_eval())) ||
+         //   (!isNull(lhs.max_eval()) && !isNull(rhs.max_eval()) && (*lhs.max_eval() == *rhs.max_eval()))) &&
+         (lhs.tolerance_grad() == rhs.tolerance_grad()) &&
+         (lhs.tolerance_change() == rhs.tolerance_change() &&
+         (lhs.history_size() == rhs.history_size()));
+}
+
+void LBFGSOptions::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lr);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(max_iter);
+  //_TORCH_OPTIM_SERIALIZE_TORCH_ARG(max_eval);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(tolerance_grad);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(tolerance_change);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(history_size);
+  //_TORCH_OPTIM_SERIALIZE_TORCH_ARG(line_search_fn);
+}
+
+void LBFGSOptions::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lr);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, max_iter);
+  //_TORCH_OPTIM_DESERIALIZE_TORCH_ARG(c10::optional<int64_t>, max_eval);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, tolerance_grad);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, tolerance_change);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, history_size);
+  //_TORCH_OPTIM_DESERIALIZE_TORCH_ARG(c10::optional<int64_t>, line_search_fn);
+}
+
+Tensor LBFGS::_gather_flat_grad() {
   std::vector<Tensor> views;
-  for (auto& parameter : parameters_) {
+  for (auto& parameter : _params) {
     if (!parameter.grad().defined()) {
-      views.push_back(parameter.new_empty({parameter.numel()}).zero_());
+      views.emplace_back(parameter.new_empty({parameter.numel()}).zero_());
```

Comment:
It might be easier to compare with Python implementation if we have a temporary variable `Tensor view` and do `view = ...` in each branch, and then do `views.push_back(view)` at the end